src/memory_provider.c

```diff
@@ -7,29 +7,67 @@
  *
  */
 
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <umf/memory_provider.h>
+
 #include "memory_provider_internal.h"
 #include "utils_common.h"
+#include "utils_concurrency.h"
 
-#include <umf/memory_provider.h>
+// base allocator used by all providers and pools
+umf_ba_pool_t *BA_pool = NULL;
 
-#include <assert.h>
-#include <stdio.h>
-#include <stdlib.h>
+// reference counter of the global base allocator
+static int N_refs_BA_pool = 0;
 
 typedef struct umf_memory_provider_t {
     umf_memory_provider_ops_t ops;
     void *provider_priv;
 } umf_memory_provider_t;
 
+int umfBaseAllocatorCreate(void) {
+    if (BA_pool == NULL) {
+        umf_ba_pool_t *ba_pool = umf_ba_create(SIZE_BA_POOL_CHUNK);
+        if (!ba_pool) {
+            return -1;
+        }
+
+        if (!util_atomic_bool_compare_exchange(&BA_pool, NULL, ba_pool)) {
+            umf_ba_destroy(ba_pool);
+        }
+    }
+
+    util_atomic_increment(&N_refs_BA_pool);
+
+    assert(BA_pool != NULL);
+    return 0;
+}
+
+void umfBaseAllocatorDestroy(void) {
+    if (util_atomic_decrement(&N_refs_BA_pool) == 0) {
```

Comment:
> There is a race possible here: consider a case where this condition is true, but just before calling `umf_ba_destroy(BA_pool);` some other thread executes `umfBaseAllocatorCreate` and sets BA_pool to that new pool.
> 
> I think you need to do compare_exchange on the BA_pool instead of just assigning NULL. Something like this:
> 
> ```
> while (1) {
>     umf_ba_pool_t *pool = util_atomic_load(&BA_pool);
>     if (pool == NULL) return;
>     if (util_atomic_decrement(&N_refs_BA_pool) == 0) {
>         if (!util_atomic_bool_compare_exchange(&BA_pool, pool , NULL)) {
>             umf_ba_destroy(pool);
>          }
>      } else { return; }
>  }
> ```

I'm afraid it is incorrect either, since `util_atomic_decrement()` can be called many times here ...

Reply:
Done